File: src/token.ts
export const OPS = ['+', '-', '*', '/'] as const;
export type Op = (typeof OPS)[number];

/** Generated token. `type` is always `'op'`; `value` is inferred from the input. */
export type Token<V extends Op = Op> = {
  readonly type: 'op';
  readonly value: V;
};

export function generateToken<V extends Op>(value: V): Token<V> {
  return { type: 'op', value };
}

export type TokenMap = { readonly [K in Op]: Token<K> };

export function generateTokens(): TokenMap {
  return {
    '+': generateToken('+'),
    '-': generateToken('-'),
    '*': generateToken('*'),
    '/': generateToken('/'),
  };
}

File: android/app/src/main/java/com/skc45/drcr2/Token.kt
package com.skc45.drcr2

data class Token(val type: String = "op", val value: String) {
    fun toJson(): String = """{"type":"$type","value":"$value"}"""
}

private val OPS = listOf("+", "-", "*", "/")

fun generateToken(value: String): Token? =
    if (value in OPS) Token(value = value) else null

fun generateTokens(): Map<String, Token> =
    OPS.associateWith { Token(value = it) }
